shotover/src/server.rs

```diff
@@ -212,11 +291,44 @@ impl<C: CodecBuilder + 'static> TcpCodecListener<C> {
                 crate::connection_span::span(self.connection_count, self.source_name.as_str());
             let transport = self.transport;
 
-            async {
+            let should_return = async {
                 tokio::select! {
-                    // Normal Connection handling
-                    stream_result = Self::accept(&mut self.listener) => {
-                        let stream = stream_result?;
+                    // Wait for a permit to become available and accept new connection
+                    stream_result = async {
+                        // Wait for a permit to become available
+                        let permit = if self.hard_connection_limit {
+                            match self.limit_connections.clone().try_acquire_owned() {
+                                Ok(p) => p,
+                                Err(_e) => {
+                                    //close the socket too full!
+                                    self.listener = None;
+                                    tokio::time::sleep(Duration::from_secs(1)).await;
+                                    return Err(anyhow!("Connection limit reached, retrying"));
+                                }
+                            }
+                        } else {
+                            self.limit_connections.clone().acquire_owned().await?
+                        };
+
+                        // Only accept new connections if socket hasn't been handed off
+                        if self.socket_handed_off {
+                            // Socket handed off, do not accept new connections
+                            futures::future::pending::<Result<(TcpStream, OwnedSemaphorePermit)>>().await
```

Comment:
This will prevent gradual shutdown from occuring if 1 second has passed since the socket was handed off.